### PR TITLE
[BEAM-1138] merging KeyedPValueTrackingVisitor with DirectGraphVisitor

### PR DESCRIPTION
**Pull request by hepeknet**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-1138] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

@kennknowles does this look good?
---

### CODE REVIEW

**Comment by kennknowles** [comment]
Thanks for putting this PR together. I think @tgroh is the best reviewer for this one.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java [hunk 1]
```diff
@@ -159,7 +159,7 @@ public DirectPipelineResult run(Pipeline pipeline) {
     pipeline.traverseTopologically(graphVisitor);
 
     @SuppressWarnings("rawtypes")
-    KeyedPValueTrackingVisitor keyedPValueVisitor = KeyedPValueTrackingVisitor.create();
+    DirectGraphVisitor keyedPValueVisitor = new DirectGraphVisitor();
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java]: Remove this line. We should just use the existing graphVisitor

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectGraphVisitorTest.java [hunk 2]
```diff
@@ -87,7 +86,7 @@ public void processElement(DoFn<String, String>.ProcessContext c)
   @Test
   public void getRootTransformsContainsRootTransforms() {
     PCollection<String> created = p.apply(Create.of("foo", "bar"));
-    PCollection<Long> counted = p.apply(Read.from(CountingSource.upTo(1234L)));
+    PCollection<Long> counted = p.apply(GenerateSequence.from(0).to(1234L));
```
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectGraphVisitorTest.java]: revert these changes.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitorTest.java [hunk 3]
```diff
@@ -54,19 +54,21 @@
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
 
-/** Tests for {@link KeyedPValueTrackingVisitor}. */
+/** Tests for {@link DirectGraphVisitor#getKeyedPValues()}. */
```
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitorTest.java]: Merge the tests into the `DirectGraphVisitorTest`

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/EvaluationContextTest.java [hunk 4]
```diff
@@ -101,8 +101,8 @@ public void setup() {
     view = created.apply(View.<Integer>asIterable());
     unbounded = p.apply(GenerateSequence.from(0));
 
-    KeyedPValueTrackingVisitor keyedPValueTrackingVisitor = KeyedPValueTrackingVisitor.create();
-    p.traverseTopologically(keyedPValueTrackingVisitor);
+    DirectGraphVisitor graphVisitor = new DirectGraphVisitor();
+    p.traverseTopologically(graphVisitor);
 
     BundleFactory bundleFactory = ImmutableListBundleFactory.create();
     graph = DirectGraphs.getGraph(p);
```
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/EvaluationContextTest.java]: We can either use the existing `DirectGraphVisitor` or add a utility method to `DirectGraphs`. I'm in favor of the former.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectGraphVisitor.java [hunk 5]
```diff
@@ -113,8 +140,38 @@ private String genStepName() {
...
     checkState(finalized, "Can't get a graph before the Pipeline has been completely traversed");
     return DirectGraph.create(producers, primitiveConsumers, views, rootTransforms, stepNames);
   }
+
+  /**
+   * Get all keyed {@link PValue PValues}. A {@link PValue} is keyed if it
+   * is the result of a {@link PTransform} that produces keyed outputs. A {@link PTransform} that
+   * produces keyed outputs is assumed to colocate output elements that share a key.
+   *
+   * <p>All {@link GroupByKey} transforms, or their runner-specific implementation primitive,
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectGraphVisitor.java]: This part of the comment can be RMed.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectGraphVisitorTest.java (hunk 2 above)
Comment (hepeknet) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectGraphVisitorTest.java]: are you sure about this? CountingSource.upTo(1234L) is deprecated

Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectGraphVisitorTest.java]: Yes. These changes aren't required for this change, and I'd like to avoid making unnecessary changes where possible.

As an aside, the deprecation doesn't concern me (`CountingSource` shouldn't be used by users as `GenerateSequence` is the abstraction we'd prefer they'd use, but it's acceptable to use in runner tests).

**Comment by kennknowles** [comment]
Looks like this PR now has conflicts - can you rebase and force push please?

**Comment by davorbonaci** [comment]
Any updates on this PR perhaps?
